lib/imagery/sig.c

```diff
@@ -140,3 +267,181 @@ int I_write_signatures(FILE * fd, struct Signature *S)
+            new_vars[c][i] = (double *)G_malloc(S->nbands * sizeof(double));
+    }
+
+    /* Obtain order of matching items */
+    for (unsigned int j = R->nfiles; j--;) {
+        for (unsigned int i = S->nbands; i--;) {
+            if (S->bandrefs[i] && group_bandrefs[j] &&
+                !strcmp(S->bandrefs[i], group_bandrefs[j])) {
+                    if (complete) {
+                        /* Reorder pointers to existing strings only */
+                        new_bandrefs[j] = S->bandrefs[i];
+                        new_order[i] = j;
+                    }
+                    /* Keep a track of matching items for error reporting */
+                    match1[i] = 1;
+                    match2[j] = 1;
+                    break;
+            }
+        }
+    }
+
+    /* Check for band reference mismatch */
+    mc1 = mc2 = 0;
+    mismatches = (char **)G_malloc(2 * sizeof(char **));
+    mismatches[0] = NULL;
+    mismatches[1] = NULL;
+    total = 1;
+    for (unsigned int i = 0; i < S->nbands; i++) {
+        if (!match1[i]) {
+            if (S->bandrefs[i])
+                total = total + strlen(S->bandrefs[i]);
+            else
+                total = total + 24;
+            mismatches[0] = (char *)G_realloc(mismatches[0], total * sizeof(char *));
+            if (mc1)
+                strcat(mismatches[0], ",");
+            else
+                mismatches[0][0] = '\0';
+            if (S->bandrefs[i])
+                strcat(mismatches[0], S->bandrefs[i]);
+            else
+                strcat(mismatches[0], "<band reference missing>");
+            mc1++;
+            total = total + 1;
+        }
+    }
+    total = 1;
+    for (unsigned int j = 0; j < R->nfiles; j++) {
+        if (!match2[j]) {
+            if (group_bandrefs[j])
+                total = total + strlen(group_bandrefs[j]);
+            else
+                total = total + 24;
+            mismatches[1] = (char *)G_realloc(mismatches[1], total * sizeof(char *));
+            if (mc2)
+                strcat(mismatches[1], ",");
+            else
+                mismatches[1][0] = '\0';
+            if (group_bandrefs[j])
+                strcat(mismatches[1], group_bandrefs[j]);
+            else
+                strcat(mismatches[1], "<band reference missing>");
+            mc2++;
+            total = total + 1;
+        }
+    }
+
+    /* Swap var matrix values in each of classes */
+    if (!mc1 && !mc2) {
+        for (unsigned int c = S->nsigs; c--;) {
+            for (unsigned int b1 = 0; b1 < S->nbands; b1++) {
+                new_means[c][new_order[b1]] = S->sig[c].mean[b1];
+                for (unsigned int b2 = 0; b2 <= b1; b2++) {
+                    if (new_order[b1] > new_order[b2]) {
+                        new_vars[c][new_order[b1]][new_order[b2]] = S->sig[c].var[b1][b2];
+                    }
+                    else {
+                        new_vars[c][new_order[b2]][new_order[b1]] = S->sig[c].var[b1][b2];
```

Comment:
Running example from i.maxlik manual:
`i.maxlik group=lsat7_2002 subgroup=lsat7_2002 signaturefile=sig_cluster_lsat2002 output=lsat7_2002_cluster_classes reject=lsat7_2002_cluster_reject`

resulted in:
```
zsh: segmentation fault  i.maxlik group=lsat7_2002 subgroup=lsat7_2002 


Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libgrass_imagery.8.0.dylib    	0x00000001088e5877 I_sort_signatures_by_bandref + 2151 (sig.c:410)
1   i.maxlik                      	0x00000001088c2470 open_files + 464 (open.c:49)
2   i.maxlik                      	0x00000001088c1dd5 main + 661 (main.c:111)
3   libdyld.dylib                 	0x00007fff6ff21cc9 start + 1

```

Reply:
The seg fault is reproducable (for me) using nc_spm_full_v2alpha2 with:

```sh
g.mapset -c testmaxlik

g.region raster=lsat7_2002_10@PERMANENT -p

g.copy raster=lsat7_2002_10@PERMANENT,lsat7_2002_10,\
lsat7_2002_20@PERMANENT,lsat7_2002_20,\
lsat7_2002_30@PERMANENT,lsat7_2002_30,\
lsat7_2002_40@PERMANENT,lsat7_2002_40,\
lsat7_2002_50@PERMANENT,lsat7_2002_50,\
lsat7_2002_70@PERMANENT,lsat7_2002_70

r.support map=lsat7_2002_10 bandref=TM7_1
r.support map=lsat7_2002_20 bandref=TM7_2
r.support map=lsat7_2002_30 bandref=TM7_3
r.support map=lsat7_2002_40 bandref=TM7_4
r.support map=lsat7_2002_50 bandref=TM7_5
r.support map=lsat7_2002_70 bandref=TM7_7

i.group group=lsat7_2002 subgroup=lsat7_2002\
    input=lsat7_2002_10,lsat7_2002_20,lsat7_2002_30,lsat7_2002_40,lsat7_2002_50,lsat7_2002_70

i.cluster group=lsat7_2002 subgroup=lsat7_2002 signaturefile=sig_cluster_lsat2002\
    classes=10 reportfile=rep_clust_lsat2002.txt

i.maxlik group=lsat7_2002 subgroup=lsat7_2002 signaturefile=sig_cluster_lsat2002\
    output=lsat7_2002_cluster_classes reject=lsat7_2002_cluster_reject
```

however, setting `i.cluster` classes <= nbands (in the example above 6 or less) it works. I cannot figure out where the problem is. (Debugging reveals that the segfault is raised when attempting to access new_vars[3][0][0].)

 I extracted the logic to a testing c file and with the equivalent changes of:

```diff
diff --git a/lib/imagery/sig.c b/lib/imagery/sig.c
index 54262f636..ec530d69f 100644
--- a/lib/imagery/sig.c
+++ b/lib/imagery/sig.c
@@ -325,7 +325,7 @@ char **I_sort_signatures_by_bandref(struct Signature *S, const struct Ref *R) {
     /* Allocate memory for temporary storage of sorted values */
     new_bandrefs = (char **)G_malloc(S->nbands * sizeof(char **));
     new_means = (double **)G_malloc(S->nbands * sizeof(double *));
-    new_vars = (double ***)G_malloc(S->nbands * sizeof(double **));
+    new_vars = (double ***)G_malloc(S->nsigs * sizeof(double **));
     for (unsigned int c = S->nsigs; c--;) {
         new_means[c] = (double *)G_malloc(S->nbands * sizeof(double));
         new_vars[c] = (double **)G_malloc(S->nbands * sizeof(double *));
```

things seem to work as expected for the test file at least (but *not* for GRASS).

Reply:
Everything is as it should be. This code was hard to write, it should be hard to read and use! (I was able to reproduce this issue locally, I'll look into this tomorrow. At first I'll add a test triggering a crash to have better coverage of this function)

Reply:
There was an error in memory allocation (as you correctly spotted already). I created a crashing test case and then fixed memory allocation – now I do not see any crashes on my side any more. Please try again to see if it is fixed on your system too.

Reply:
No feedback in two weeks.